### 1.- Given the lyrics you selected in your previous homework, compute their similarity.

In [1]:
import nltk
from nltk import word_tokenize, pos_tag, ne_chunk

In [3]:
myfile1 = open('/Users/sergi/Desktop/lyrics/daydreaming.txt', 'r')
lyric1 = myfile1.read().replace('\n', " ")

myfile2 = open('/Users/sergi/Desktop/lyrics/declareindependence.txt', 'r')
lyric2 = myfile2.read().replace('\n', " ")

myfile3 = open('/Users/sergi/Desktop/lyrics/dontworry.txt', 'r')
lyric3 = myfile3.read().replace('\n', " ")

myfile4 = open('/Users/sergi/Desktop/lyrics/everythingnow.txt', 'r')
lyric4 = myfile4.read().replace('\n', " ")

myfile5 = open('/Users/sergi/Desktop/lyrics/foreveryoung.txt', 'r')
lyric5 = myfile5.read().replace('\n', " ")

myfile6 = open('/Users/sergi/Desktop/lyrics/gyalyouapartyanimal.txt', 'r')
lyric6 = myfile6.read().replace('\n', " ")

myfile7 = open('/Users/sergi/Desktop/lyrics/ifeelitcoming.txt', 'r')
lyric7 = myfile7.read().replace('\n', " ")

myfile8 = open('/Users/sergi/Desktop/lyrics/isthisit.txt', 'r')
lyric8 = myfile8.read().replace('\n', " ")

myfile9 = open('/Users/sergi/Desktop/lyrics/shapeofyou.txt', 'r')
lyric9 = myfile9.read().replace('\n', " ")

myfile10 = open('/Users/sergi/Desktop/lyrics/wonderfullife.txt', 'r')
lyric10 = myfile10.read().replace('\n', " ")

In [5]:
from __future__ import division
import string
import math

tokenize = lambda doc: doc.lower().split(" ")

all_lyrics = [lyric1, lyric2, lyric3, lyric4, lyric5, lyric6, lyric7, lyric8, lyric9, lyric10]

tokenized_lyrics = [tokenize(d) for d in all_lyrics] # tokenized docs
all_tokens_set = set([item for sublist in tokenized_lyrics for item in sublist])

from sklearn.feature_extraction.text import TfidfVectorizer

sklearn_tfidf = TfidfVectorizer(norm='l2',min_df=0, use_idf=True, smooth_idf=False, sublinear_tf=True, tokenizer=tokenize)

sklearn_representation = sklearn_tfidf.fit_transform(all_lyrics)
print(sklearn_representation.toarray()[0].tolist())
print(lyric1)

[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.05482327768845429, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.06066503879653599, 0.0, 0.0, 0.0, 0.0, 0.0749333918999829, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.34375361782928576, 0.16380044074146252, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.1093118592149584, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.16380044074146252, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.27733825441588367, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.27733825441588367, 0.0, 0.16380044074146252, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0

In [6]:
def cosine_similarity(vector1, vector2):
    dot_product = sum(p*q for p,q in zip(vector1, vector2))
    magnitude = math.sqrt(sum([val**2 for val in vector1])) * math.sqrt(sum([val**2 for val in vector2]))
    if not magnitude:
        return 0
    return dot_product/magnitude

In [7]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[1])

0.016801409771555947

In [8]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[2])

0.101367517268961

In [9]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[3])

0.081488114815315441

In [10]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[4])

0.084041209545207732

In [11]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[5])

0.076718860850278292

In [12]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[6])

0.08708007888451634

In [13]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[7])

0.086206141380348822

In [14]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[8])

0.074951576464716702

In [15]:
cosine_similarity(sklearn_representation.toarray()[0],sklearn_representation.toarray()[9])

0.11102894448538542

In [17]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[2])

0.10517359005652686

In [18]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[3])

0.03059412440055825

In [19]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[4])

0.082769062415530462

In [20]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[5])

0.032495043752907418

In [21]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[6])

0.05520477571287908

In [22]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[7])

0.067245902958047152

In [23]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[8])

0.065352328729234244

In [24]:
cosine_similarity(sklearn_representation.toarray()[1],sklearn_representation.toarray()[9])

0.05376566609040629

In [26]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[3])

0.10877734064572017

In [27]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[4])

0.18290002515354725

In [28]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[5])

0.12842258074907142

In [29]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[6])

0.1765214389008003

In [30]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[7])

0.10478173753009655

In [31]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[8])

0.15469683142202165

In [32]:
cosine_similarity(sklearn_representation.toarray()[2],sklearn_representation.toarray()[9])

0.16258914427395382

In [34]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[4])

0.11411222227913445

In [35]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[5])

0.073623408230288956

In [36]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[6])

0.13496278275477275

In [37]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[7])

0.11907813154137763

In [38]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[8])

0.15977265113692857

In [39]:
cosine_similarity(sklearn_representation.toarray()[3],sklearn_representation.toarray()[9])

0.14945886336978273

In [40]:
cosine_similarity(sklearn_representation.toarray()[4],sklearn_representation.toarray()[5])

0.086599289768759191

In [41]:
cosine_similarity(sklearn_representation.toarray()[4],sklearn_representation.toarray()[6])

0.12062712095773982

In [42]:
cosine_similarity(sklearn_representation.toarray()[4],sklearn_representation.toarray()[7])

0.10110789674850014

In [43]:
cosine_similarity(sklearn_representation.toarray()[4],sklearn_representation.toarray()[8])

0.14422164190767492

In [44]:
cosine_similarity(sklearn_representation.toarray()[4],sklearn_representation.toarray()[9])

0.095012654853919742

In [45]:
cosine_similarity(sklearn_representation.toarray()[5],sklearn_representation.toarray()[6])

0.111829958888176

In [46]:
cosine_similarity(sklearn_representation.toarray()[5],sklearn_representation.toarray()[7])

0.08239416943421736

In [47]:
cosine_similarity(sklearn_representation.toarray()[5],sklearn_representation.toarray()[8])

0.16988378876501195

In [48]:
cosine_similarity(sklearn_representation.toarray()[5],sklearn_representation.toarray()[9])

0.080279557379962757

In [49]:
cosine_similarity(sklearn_representation.toarray()[6],sklearn_representation.toarray()[7])

0.13929925322947884

In [50]:
cosine_similarity(sklearn_representation.toarray()[6],sklearn_representation.toarray()[8])

0.15070516894832967

In [51]:
cosine_similarity(sklearn_representation.toarray()[6],sklearn_representation.toarray()[9])

0.10974459110166558

In [52]:
cosine_similarity(sklearn_representation.toarray()[7],sklearn_representation.toarray()[8])

0.15291166903412526

In [53]:
cosine_similarity(sklearn_representation.toarray()[7],sklearn_representation.toarray()[9])

0.098511902819696462

In [54]:
cosine_similarity(sklearn_representation.toarray()[8],sklearn_representation.toarray()[9])

0.12445041387863318

### The most similar lyrics are lyric3 and lyric5 with a cosine similarity of 0.1829

In [59]:
all_lyrics_names = ['lyric1: daydreaming', 'lyric2: declare independence', 'lyric3: dontworry', 'lyric4: everything now', 'lyric5: foreveryoung', 'lyric6: gyal you a party animal', 'lyric7: i feel it coming', 'lyric8: is this it', 'lyric9: shapeofyou', 'lyric10: wonderfullife']

for i in range(10):
    for j in range(i+1,10):
        similarity = cosine_similarity(sklearn_representation.toarray()[i], sklearn_representation.toarray()[j])
        print("The similarity between the {} and the {} is {}\n\n".format(all_lyrics_names[i], all_lyrics_names[j], similarity))

The similarity between the lyric1: daydreaming and the lyric2: declare independence is 0.016801409771555947


The similarity between the lyric1: daydreaming and the lyric3: dontworry is 0.101367517268961


The similarity between the lyric1: daydreaming and the lyric4: everything now is 0.08148811481531544


The similarity between the lyric1: daydreaming and the lyric5: foreveryoung is 0.08404120954520773


The similarity between the lyric1: daydreaming and the lyric6: gyal you a party animal is 0.07671886085027829


The similarity between the lyric1: daydreaming and the lyric7: i feel it coming is 0.08708007888451634


The similarity between the lyric1: daydreaming and the lyric8: is this it is 0.08620614138034882


The similarity between the lyric1: daydreaming and the lyric9: shapeofyou is 0.0749515764647167


The similarity between the lyric1: daydreaming and the lyric10: wonderfullife is 0.11102894448538542


The similarity between the lyric2: declare independence and the lyric3: d

### 2.- Improve the exercise by first removing the stop words and by stemming the lyrics. Recompute the similarity. Did anything change, apart from the values (e.g., the two most similar songs changed)? Write a markdown with some comments.

In [62]:
from nltk.corpus import stopwords

In [76]:
all_lyrics_new = []
porter = nltk.PorterStemmer()

for lyric in tokenized_lyrics:
    all_lyrics_swremoved = [w for w in lyric if w not in stopwords.words('english')]
    print (all_lyrics_swremoved,'\n')
    all_lyrics_stemming = [porter.stem(t) for t in all_lyrics_swremoved]
    print (all_lyrics_stemming, '\n')
    all_lyrics_new.append(all_lyrics_stemming)
    print (all_lyrics_new, '\n')

['dreamers', 'never', 'learn', 'never', 'learn', 'beyond,', 'beyond', 'point', 'return', 'return', "it's", 'late', 'damage', 'done', 'damage', 'done', 'goes', 'beyond', 'beyond', 'white', 'room', 'window', 'sun', 'comes', 'happy', 'serve', 'happy', 'serve'] 

['dreamer', 'never', 'learn', 'never', 'learn', 'beyond,', 'beyond', 'point', 'return', 'return', "it'", 'late', 'damag', 'done', 'damag', 'done', 'goe', 'beyond', 'beyond', 'white', 'room', 'window', 'sun', 'come', 'happi', 'serv', 'happi', 'serv'] 

[['dreamer', 'never', 'learn', 'never', 'learn', 'beyond,', 'beyond', 'point', 'return', 'return', "it'", 'late', 'damag', 'done', 'damag', 'done', 'goe', 'beyond', 'beyond', 'white', 'room', 'window', 'sun', 'come', 'happi', 'serv', 'happi', 'serv']] 

['declare', 'independence!', "don't", 'let', 'you!', 'declare', 'independence!', "don't", 'let', 'you!', '(justice)', 'declare', 'independence!', "don't", 'let', 'you!', 'declare', 'independence!', "don't", 'let', 'you!', '(justice)',

['every', 'inch', "sky's", 'got', 'star', 'every', 'inch', "skin's", 'got', 'scar', 'guess', "you've", 'got', 'everything', 'every', 'inch', 'space', 'head', 'filled', 'things', 'read', 'guess', "you've", 'got', 'everything', 'every', 'film', "you've", 'ever', 'seen', 'fills', 'spaces', 'dreams', 'reminds', '(everything', 'everything', 'now)', 'every', 'inch', "road's", 'got', 'sign', 'every', 'boy', 'uses', 'line', 'pledge', 'allegiance', 'everything', 'every', 'song', "i've", 'ever', 'heard', 'playing', 'time,', "it's", 'absurd', 'reminds', 'me,', "we've", 'got', 'everything', 'turn', 'speakers', 'till', 'break', "'cause", 'every', 'time', 'smile', "it's", 'fake!', 'stop', 'pretending,', "you've", 'got', 'need', '(everything', 'now!)', 'want', '(everything', 'now!)', "can't", 'live', 'without', '(everything', 'now!)', "can't", 'live', 'without', '(everything', 'now!)', '(everything', 'now!)', 'everything', 'every', 'ancient', "road's", 'got', 'town', 'daddy,', 'come', "you're", 'neve

['yeah,', 'uh', 'tell', 'really', 'like', 'baby', 'take', 'time', "don't", 'ever', 'fight', 'take', 'step-by-step', 'see', 'eyes', "'cause", 'never', 'tell', 'lies', 'feel', 'body', 'shake', 'heat', 'legs', "you've", 'scared', 'love', "don't", 'run,', 'know', "you've", 'simple', 'touch', 'set', 'free', "don't", 'rush', "you're", 'alone', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'single', 'type', 'baby,', 'perfect', 'time', "i'm", 'trying', 'get', 'high', 'faded', 'touch', "don't", 'need', 'lonely', 'night', 'baby,', 'make', 'right', 'got', 'let', 'try', 'give', 'want', "you've", 'scared', 'love', "don't", 'run,', 'know', "you've", 'simple', 'touch', 'set', 'free', "don't", 'rush', "you're", 'alone', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'coming,', 'feel', 'coming,', 'babe', 'feel', 'c

['go', 'sea', 'sunshine', 'fills', 'hair', 'dreams', 'hang', 'air', 'gulls', 'sky', 'blue', 'eyes', 'know', 'feels', 'unfair', "there's", 'magic', 'everywhere', 'look', 'standing', 'straight', 'sunshine', 'need', 'run', 'hide', "it's", 'wonderful,', 'wonderful', 'life', 'need', 'laugh', 'cry', "it's", 'wonderful,', 'wonderful', 'life', 'sun', 'eyes', 'heat', 'hair', 'seem', 'hate', "you're", 'need', 'friend', 'oh,', 'need', 'friend', 'make', 'happy', 'stand', 'look', 'standing', 'straight', 'sunshine'] 

['go', 'sea', 'sunshin', 'fill', 'hair', 'dream', 'hang', 'air', 'gull', 'sky', 'blue', 'eye', 'know', 'feel', 'unfair', "there'", 'magic', 'everywher', 'look', 'stand', 'straight', 'sunshin', 'need', 'run', 'hide', "it'", 'wonderful,', 'wonder', 'life', 'need', 'laugh', 'cri', "it'", 'wonderful,', 'wonder', 'life', 'sun', 'eye', 'heat', 'hair', 'seem', 'hate', "you'r", 'need', 'friend', 'oh,', 'need', 'friend', 'make', 'happi', 'stand', 'look', 'stand', 'straight', 'sunshin'] 

[['dre

In [69]:
all_lyrics2 = [' '.join(lyric) for lyric in all_lyrics_new]

In [78]:
all_lyrics2

["dreamer never learn never learn beyond, beyond point return return it' late damag done damag done goe beyond beyond white room window sun come happi serv happi serv",
 "declar independence! don't let you! declar independence! don't let you! (justice) declar independence! don't let you! declar independence! don't let you! (justice) start currency! make stamp protect languag (justice) declar independ don't let declar independ don't let",
 "oh, night don't worri 'bout thing don't worri 'bout thing don't worri 'bout thing know we'll alright don't worri 'bout thing don't worri 'bout thing don't worri 'bout thing i'll take futur forget past keep secret swear won't ask let go troubles, don't care you'v thing matter night end bright big light shine us beat tight make wanna get get like there' tomorrow like there' tomorrow like there' tomorrow oh, night don't worri 'bout thing don't worri 'bout thing don't worri 'bout thing know we'll alright don't worri 'bout thing don't worri 'bout thing do

In [70]:
sklearn_tfidf = TfidfVectorizer(norm='l2',min_df=0, use_idf=True, smooth_idf=False, sublinear_tf=True, tokenizer=tokenize)

sklearn_representation = sklearn_tfidf.fit_transform(all_lyrics2)

In [71]:
all_lyrics_names = ['lyric1: daydreaming', 'lyric2: declare independence', 'lyric3: dontworry', 'lyric4: everything now', 'lyric5: foreveryoung', 'lyric6: gyal you a party animal', 'lyric7: i feel it coming', 'lyric8: is this it', 'lyric9: shapeofyou', 'lyric10: wonderfullife']

for i in range(10):
    for j in range(i+1,10):
        similarity = cosine_similarity(sklearn_representation.toarray()[i], sklearn_representation.toarray()[j])
        print("The similarity between the {} and the {} is {}\n\n".format(all_lyrics_names[i], all_lyrics_names[j], similarity))

The similarity between the lyric1: daydreaming and the lyric2: declare independence is 0.0


The similarity between the lyric1: daydreaming and the lyric3: dontworry is 0.010005402997301165


The similarity between the lyric1: daydreaming and the lyric4: everything now is 0.03995133922008919


The similarity between the lyric1: daydreaming and the lyric5: foreveryoung is 0.020847117742723024


The similarity between the lyric1: daydreaming and the lyric6: gyal you a party animal is 0.008754584505596337


The similarity between the lyric1: daydreaming and the lyric7: i feel it coming is 0.00710431707079562


The similarity between the lyric1: daydreaming and the lyric8: is this it is 0.009913137404701056


The similarity between the lyric1: daydreaming and the lyric9: shapeofyou is 0.023902770083868253


The similarity between the lyric1: daydreaming and the lyric10: wonderfullife is 0.04769505823950047


The similarity between the lyric2: declare independence and the lyric3: dontworry 

#### The pair of lyrics most similar are the same (3 and 5), but with a low cosine similarity.  Probably, removing the stop words have made the cosine similarity to drop down.

### 3.- Run the “Word association” task on the treebank corpus. Find the most co-occurrent word of the following: 
### chairman, profit, president, investors, city, chief.

### Which considerations can you make about “chief”?

In [82]:
from nltk.corpus import treebank, stopwords
from nltk import FreqDist, ConditionalFreqDist

# initialize a new conditional distribution
cfd = ConditionalFreqDist()

# get a list of English stopwords
stopwords_list = stopwords.words('english')

def is_noun(tag):
    return tag.lower() in ['nn','nns','nn$','nn-tl','nn+bez', 'nn+hvz', 
                           'nns$','np','np$','np+bez','nps', 'nps$','nr',
                           'np-tl','nrs','nr$']

for sentence in treebank.tagged_sents():
    for (index, tagtuple) in enumerate(sentence):
        (token, tag) = tagtuple
        token = token.lower()
        if token not in stopwords_list and is_noun(tag):
            window = sentence[index+1:index+5]
            for (window_token, window_tag) in window:
                window_token = window_token.lower()
                if window_token not in stopwords_list and is_noun(window_tag):
                    cfd[token][window_token] += 1

In [83]:
print(cfd['chairman'].max())

chief


In [84]:
print(cfd['profit'].max())

%


In [85]:
print(cfd['president'].max())

chief


In [86]:
print(cfd['investors'].max())

bonds


In [87]:
print(cfd['city'].max())

streets


In [88]:
print(cfd['chief'].max())

officer


### We can see that the word "chief" is used in similar contexts where there is a responsibility of importance.